### ISSUE TITLE
[RFC] DynamicProperty enhancements

### ISSUE THREAD

**Issue by NachoSoto** [ReactiveCocoa/ReactiveCocoa#2535]
@andymatuschak and I were exploring ways to use `DynamicProperty` more, instead of doing ad-hoc subscriptions.

The particular example we were working with is the typical (in RAC 2 terms) `RAC(imageView, image) = signal`. Turns out that’s awfully hard to do with `RAC 4`: there’s no way to create an extension of `UIImageView` that provides a `MutableProperty` that allows us to do 2 things:

- Guarantee that subscription of a binded producer (or signal) happens on the main thread.
- Change the type of the `DynamicProperty` to `UIImage`.

#### More specifically:
- It’s currently impossible to compose a `MutableProperty`. But even if it were possible (with something like `AnyMutableProperty`), it’s impossible to "make changes” to a signal or producer that gets binded to it.
- `DynamicProperty` is implemented with `Value` `AnyObject?`.

#### Suggested improvements:
- Create some sort of `QueueDispatchingMutableProperty` (see an example below) that dispatches setting the value to a particular queue.
- Change `DynamicProperty` to be generic over `Value`, and doing `as! Value`: this would require removing its `@objc` qualifier, which is a breaking change, but I really can't imagine why you'd want to use `DynamicProperty` from `Obj-C`.

I feel very strongly about the second point, I think that would be a positive change. But I'm not sure at all about the first one. Ideally, instead of something so specific, `MutablePropertyType` could be more easily composed to introduce modifications like this one, or like #2431.

Of course we could just implement this in our own codebase, but I'm thinking about how people can accomplish this using `RAC 4`:

```swift
let someImageProducer: SignalProducer<UIImage?, NoError>
someImageView.imageProperty <~ someImageProducer
```

And ideally this is as easy to implement as this:

```swift
extension UIImageView {
	public var imageProperty: MutablePropertyType<UIImage?> {
		return QueueDispatchingMutableProperty(
			property: DynamicProperty<UIImage?>(object: self, keyPath: "image"),
			dispatchQueue: dispatch_get_main_queue()
		)
	}
}
```

Any thoughts on this appreciated! :star: :heart: 

---------------
Example of something like `QueueDispatchingMutableProperty`:
```swift
public final class QueueDispatchingMutableProperty<Value>: MutablePropertyType {
	private let _valueGetter: () -> Value
	private let _valueSetter: Value -> ()
	private let _producer: () -> SignalProducer<Value, NoError>
	private let _dispatchQueue: dispatch_queue_t

	public var value: Value {
		get { return _valueGetter() }
		set { dispatch_async(_dispatchQueue) { self._valueSetter(newValue) } }
	}

	public var producer: SignalProducer<Value, NoError> {
		return _producer()
	}

	/// Initializes a property as a view of the given property.
	public init<P: MutablePropertyType where P.Value == Value>(property: P, dispatchQueue: dispatch_queue_t) {
		_valueGetter = { property.value }
		_valueSetter = { value in property.value = value }
		_producer = { property.producer }
		_dispatchQueue = dispatchQueue
	}
}
```

**Comment by neilpa** [comment]
I'm :-1: on hiding an implicit dispatch behind a property binding. An `assert` or `precondition` if you're on the wrong thread feels much more appropriate.

**Comment by neilpa** [comment]
Also, I've got similar ideas in [Rex](https://github.com/neilpa/Rex/blob/b4f891e4b698c07268d5990c1aeb5f66b9523f0f/Source/UIKit/UIImageView.swift) already. I've just been holding off on porting them cause I wanted a bit more baking time. (And on the hope that we get an open source Swift compiler soon).

**Comment by jspahrsummers** [comment]
Pretty sure `@objc` is necessary for KVO to work. Are you proposing removing KVO support there?

(No strong opinion either way, just think it should be called out.)

**Comment by NachoSoto** [comment]
Thanks for the comments guys!

> An assert or precondition if you're on the wrong thread feels much more appropriate.

I'm looking for ways to guarantee this **at complete time** though.

> I'm :-1: on hiding an implicit dispatch behind a property binding.

In any case, I'm interested in the broader conversation: how to be able to compose properties to do these kinds of things, whether it's adding an `assert`ion, `filter`ing values, `map`ping values, etc.

> Also, I've got similar ideas in Rex already.

Nice! Yeah, I think something like `rex_valueProperty` is exactly what we want here: being able to create a `MutableProperty<T>` binded to a Cocoa property. However, I don't think we need another abstraction to accomplish this. I think my suggested changes to `DynamicProperty` would accomplish this.

> Pretty sure @objc is necessary for KVO to work. Are you proposing removing KVO support there?

We only need KVO on the target `object` though (which we get using `object.rac_valuesForKeyPath`), not on the class itself, right? Or am I missing something?

**Comment by neilpa** [comment]
>I'm looking for ways to guarantee this at ~~complete~~ compile time though.

That would be ideal. However, there's plenty of times where the signal is already on the main thread. Hiding this behind properties will introduce an extra scheduling hop which is expensive. Even if we use the `mainThreadScheduler` to short circuit the hop, that still requires wrapping every bound producer. I've run into numerous performance issues in past iOS projects when trying implicitly hide these hops so want to be careful here.

The "right way" to do this is embedding the scheduler in the signal's type. Sadly, I've not figured out a way to achieve this that isn't ugly and cumbersome to use.

> In any case, I'm interested in the broader conversation: how to be able to compose properties to do these kinds of things, whether it's adding an assertion, filtering values, mapping values, etc.

Isn't this just `Property -> Producer -> ... -> Property`? We could explore making this easier by having `PropertyType` conform to `SignalProducerType` and then make another variant of `lift` for properties.

As an aside, I've also been thinking about other ways to [compose properties as logical predicates](https://github.com/neilpa/Rex/pull/48). It's pretty rough since I haven't had the time to flush out my ideas. The extension though is that properties should compose in much the same way as mathematical operators since they provide a current value.

> Nice! Yeah, I think something like rex_valueProperty is exactly what we want here: being able to create a MutableProperty<T> binded to a Cocoa property. However, I don't think we need another abstraction to accomplish this. I think my suggested changes to DynamicProperty would accomplish this.

I disagree. `DynamicProperty` exists as the raw runtime bridge between Objective-C and Swift. The wrappers I've built are the compile-time variant of that bridge. Depending on the consumers scenario you may need one or the other. My goal was to avoid conflating the two concepts.

**Comment by DavdRoman** [comment]
The idea of making `DynamicProperty` generic sounds good to me. I've been struggling __a lot__ trying to get a simple binding between a `MutableProperty` and a `DynamicProperty` to work, and I'm certain I'm not the only one (#2425, #2198...).

About this:

> I disagree. DynamicProperty exists as the raw runtime bridge between Objective-C and Swift. The wrappers I've built are the compile-time variant of that bridge. Depending on the consumers scenario you may need one or the other. My goal was to avoid conflating the two concepts.

IMHO, I think it's worth sacrificing `DynamicProperty`'s purity in order to bring some convenience to the Swift API or, at least, building another more compiler-friendly abstraction.

**Comment by neilpa** [comment]
>> Nice! Yeah, I think something like rex_valueProperty is exactly what we want here: being able to create a MutableProperty binded to a Cocoa property. However, I don't think we need another abstraction to accomplish this. I think my suggested changes to DynamicProperty would accomplish this.

> I disagree. DynamicProperty exists as the raw runtime bridge between Objective-C and Swift. The wrappers I've built are the compile-time variant of that bridge. Depending on the consumers scenario you may need one or the other. My goal was to avoid conflating the two concepts.

Thinking further about this I do agree that the changes to better type `DynamicProperty` would be beneficial. However, I still wouldn't want to use that as the basis for creating UIKit and AppKit extension properties. Most of the backing fields don't support KVO so there's no point setting up the observation that `DynamicProperty` does by default. The best we can do is effectively property sinks which we can already get with `MutableProperty`.

**Comment by DavdRoman** [comment]
> However, I still wouldn't want to use that as the basis for creating UIKit and AppKit extension properties.

You mean not using `DynamicProperty` for UIKit/AppKit extension properties at all? Not even on properties supporting KVO?

> The best we can do is effectively property sinks which we can already get with MutableProperty.

Agreed, but in my opinion properties that do support KVO should be wrapped as `DynamicProperty`'s. It's safer and overall more reactive to non-`MutableProperty`-driven changes.

**Comment by neilpa** [comment]
> You mean not using DynamicProperty for UIKit/AppKit extension properties at all? Not even on properties supporting KVO?

Yes, because it's very ad-hoc. Also, these are almost always terminal properties so you're paying for observations that no one is listening to.

> Agreed, but in my opinion properties that do support KVO should be wrapped as DynamicProperty's. It's safer and overall more reactive to non-MutableProperty-driven changes.

In those cases you can explicitly create a `DynamicProperty`. The normal pattern is `view.prop <~ viewModel.prop`, not `view.prop <~ otherView.prop` so supporting chaining of properties directly from views shouldn't be necessary.

**Comment by DavdRoman** [comment]
> In those cases you can explicitly create a DynamicProperty.

You're totally right, I don't know what I was thinking ¯\_(ツ)_/¯

**Comment by NachoSoto** [comment]
> However, I still wouldn't want to use that as the basis for creating UIKit and AppKit extension properties. Most of the backing fields don't support KVO so there's no point setting up the observation that DynamicProperty does by default.

I agree. I wonder if we can improve `DynamicProperty` for **setting** values without encouraging the KVO aspect of it (`WriteOnlyPropertyType`?), because like you said, and the only reason why I brought this up, is to make it easier to do things like
```swift
imageView.image <~ producer (or property)
```

**Comment by NachoSoto** [comment]
Removing from 4.0, I think this can be improved after the initial 4.0 release.

**Comment by andersio** [comment]
What about a `<~` overload that takes a tuple to set up a simple observer with just `setValue:forKey:`?
```
(cocoaObject, "keyPath") <~ producer
```

**Comment by andersio** [comment]
All addressed in `BindingTarget` in RAS 1.0 and `DynamicProperty` in RAC 5.0. :)
